File: leetcode/1652/1652.go
package main

func decrypt(code []int, k int) []int {
	n := len(code)
	res := make([]int, n)

	if k == 0 {
		return res
	} else if k < 0 {
		p := -k
		window := 0

		for _, v := range code[n-p:] {
			window += v
		}
		res[0] = window

		for i := 1; i < n; i++ {
			window = window - code[(i-1-p+n)%n] + code[(i-1+n)%n]
			res[i] = window
		}
	} else {
		window := 0

		for _, v := range code[1 : k+1] {
			window += v
		}
		res[0] = window

		for i := 1; i < n; i++ {
			window = window - code[i] + code[(i+k)%n]
			res[i] = window
		}
	}
	return res
}

File: leetcode/1652/1652.py
from typing import List


class Solution:
		# sliding window
    def decrypt(self, code: List[int], k: int) -> List[int]:
        n = len(code)
        res = []
        if k == 0:
            return [0] * n
        elif k < 0:
            p = -k
            window = sum(code[i] for i in range(n-p, n))
            res.append(window)
            for i in range(1, n):
                window += -code[(i-1-p)%n] + code[(i-1)%n]
                res.append(window)
                
        elif k > 0:
            window = sum(code[i] for i in range(1, k+1))
            res.append(window)
            for i in range(1, n):
                window += -code[i] + code[(i+k)%n]
                res.append(window)
        return res